sstables/sstable_set.hh

```diff
@@ -24,84 +24,207 @@
+    // For modification methods set "capturing" to false to edit the last version
+    // of the set even if it's referenced elsewhere, leave true otherwise
+    void insert(shared_sstable sst, bool capturing = true);
+    void erase(shared_sstable sst, bool capturing = true);
+    void update(const std::vector<sstables::shared_sstable>& new_sstables,
+                const std::vector<sstables::shared_sstable>& old_sstables,
+                bool capturing = true); // strong exception guarantee
 
     // Used to incrementally select sstables from sstable set using ring-position.
     // sstable set must be alive during the lifetime of the selector.
-    class incremental_selector {
-        std::unique_ptr<incremental_selector_impl> _impl;
-        dht::ring_position_comparator _cmp;
-        mutable std::optional<dht::partition_range> _current_range;
-        mutable std::optional<nonwrapping_range<dht::ring_position_view>> _current_range_view;
-        mutable std::vector<shared_sstable> _current_sstables;
-        mutable dht::ring_position_view _current_next_position = dht::ring_position_view::min();
-    public:
-        ~incremental_selector();
-        incremental_selector(std::unique_ptr<incremental_selector_impl> impl, const schema& s);
-        incremental_selector(incremental_selector&&) noexcept;
-
-        struct selection {
-            const std::vector<shared_sstable>& sstables;
-            dht::ring_position_view next_position;
-        };
-
-        // Return the sstables that intersect with `pos` and the next
-        // position where the intersecting sstables change.
-        // To walk through the token range incrementally call `select()`
-        // with `dht::ring_position_view::min()` and then pass back the
-        // returned `next_position` on each next call until
-        // `next_position` becomes `dht::ring_position::max()`.
-        //
-        // Successive calls to `select()' have to pass weakly monotonic
-        // positions (incrementability).
-        //
-        // NOTE: both `selection.sstables` and `selection.next_position`
-        // are only guaranteed to be valid until the next call to
-        // `select()`.
-        selection select(const dht::ring_position_view& pos) const;
+    class incremental_selector;
+    incremental_selector make_incremental_selector(uint64_t version) const;
+    incremental_selector make_incremental_selector() const;
+    friend class version_reference;
+};
+
+lw_shared_ptr<sstable_set> make_partitioned_sstable_set(schema_ptr schema, bool use_level_metadata = true);
+
+class version_reference : public weakly_referencable<version_reference>, public enable_lw_shared_from_this<version_reference> {
+    lw_shared_ptr<sstable_set> _versioned_ssts;
+    uint64_t _version;
+public:
+    version_reference(lw_shared_ptr<sstable_set> versioned_ssts, uint64_t version);
+    ~version_reference();
+    uint64_t get() const;
+};
+
+class sstable_set::incremental_selector {
```

Comment:
I think I needed to move it when I was still working on it to avoid working on incomplete types (it had a version_reference which had a lw_shared to the sstable_set). Looks like in the final version I can move it back